src/Nixfmt/Predoc.hs

```diff
@@ -89,44 +81,46 @@ instance Pretty a => Pretty (Maybe a) where
     pretty Nothing  = mempty
     pretty (Just x) = pretty x
 
-instance Semigroup (Tree a) where
-    left <> right = Node left right
+instance Semigroup Doc where
+    x <> (Node NoAnn xs) = Node NoAnn (x:xs)
+    left <> right = Node NoAnn [left, right]
 
-instance Monoid (Tree a) where
-    mempty = EmptyTree
+instance Monoid Doc where
+    mempty = Node NoAnn []
+    mconcat = Node NoAnn
```

Comment:
1. This `mconcat` differs from `<>` because it doesn't consider `Node NoAnn xs` special case. Even if it was the same, it's better to write `mconcat = (<>)` to make it clear.
2. Even if you write `mconcat = (<>)` it won't be a monoid, because `mempty <> x = Node NoAnn [mempty, x] ≠ x` for all `x ≠ mempty`. So you should also explicitly consider `Node NoAnn xs` case in the `Semigroup` instance definiton.